### ISSUE TITLE
Multiple Field with dataloaders in same ObjectType

### ISSUE THREAD

**Issue by grean** [graphql-dotnet/graphql-dotnet#2266]
How to handle multiple dataloaders and how can i achieve this scenario ?
```c#
 public class AlimentType : ObjectGraphType<Aliment>
    {
        public AlimentType(IRepository repository, IDataLoaderContextAccessor accessor)
        {
            Field(i => i.Id);

            Field<AlimentGroupType, AlimentGroup>()
                .Name("alimentGroup")
                .ResolveAsync(ctx =>
                {
                    var loader = accessor.Context.GetOrAddBatchLoader<Guid, AlimentGroup>("GetAlimentGroupsById", repository.GetAlimentGroupsById);
                    return loader.LoadAsync(ctx.Source.AlimentGroupId ?? Guid.Empty);
                });
            Field(i => i.AlimentGroupId, type: typeof(GuidGraphType));

            Field<PatientType, Patient>()
                .Name("patient")
                .ResolveAsync(ctx =>
                {
                    var loader = accessor.Context.GetOrAddBatchLoader<Guid, Patient>("GetPatientsById", repository.GetPatientsById);
                    return loader.LoadAsync(ctx.Source.PatientId ?? Guid.Empty);
                });
            Field(i => i.PatientId, type: typeof(GuidGraphType));
   }
}
```
my graphQL query 
```graphql
query Aliments{
  aliments{
    id
    name
    patient{
      id
      dateOfBirth
    }
    alimentGroup{
      id
      name
    }
  }
}
```
It appears that i have missing objects sometimes, i understand that it´s because of the scoped lifetime of my EF context but is it possible to have a sequential behavior with parallel dataloaders ?

**Comment by sungam3r** [comment]
@Shane32 ?

**Comment by Shane32** [comment]
Are you asking if the data loaders can run in series while using a parallel execution strategy?

**Comment by Shane32** [comment]
(To eliminate multithreaded problems with EF)

**Comment by grean** [comment]
> Are you asking if the data loaders can run in series while using a parallel execution strategy?

Yes.
My aliment object wants its owner (a patient) and its group (alimentGroup)
So i have two navigation properties 'patient' and 'alimentGroup'

I'have a REST api with webApi, EF, autoMapper which can return a json with such a structure (my exemple graphql query). In my REST api, I use a single Linq query like this :
```c#
mydbContext.aliments.include(a => a.patient).include(a => a.alimentGroup).Tolist()
```
I simply try to mimic this behavior.

Maybe graphql philosophy is different ?

**Comment by Shane32** [comment]
Ok so graphql's dataloader won't combine both queries into a single include statement for you.  The best you're going to get with the dataloaders is something equivalent to this:
```cs
var aliments = myDbContext.Aliments.ToList();
// return graphs of the aliments
// the data loaders collect all the ids that need collecting at the second step
var patientIds = aliments.Select(x => x.PatientId).ToList();
var alimentGroupIds = aliments.Select(x => x.AlimentGroupId).ToList();

// process patients
// note that the patient ids are passed as literals to SQL, not parameters,
//   due to SQL's syntax limitations, so the query execution can't be cached within SQL
var patients = myDbContext.Patients.Select(x => patientIds.Contains(x.Id)).ToList()
// the patients are return to the proper child graphs

// process alimentgroups
var alimentGroups = myDbContext.AlimentGroups.Select(x => alimentGroupIds.Contains(x.Id)).ToList()
// the aliment groups are return to the proper child graphs
```

Note that within the preceding example, processing patients and processing alimentgroups happens simulatenously. This is a problem (as you've already discovered) with EF's single-threaded limitation.  There a several workarounds:

1. Use serial execution strategy. Since anything that does not `await` runs single-threaded anyway, it ends up that if all of your `await` statements are used on EF, and all those need to run in series, then there's nothing else that would run multi-threaded anyway.  If that's confusing, consider this example: `Field(x => x.Name)` and `Field(x => x.Company)` never run in parallel because neither has an `await` inside it.  So a serial vs parallel execution strategy don't make any difference for non-async fields.

2. Use locks on your code. Generally I wouldn't go this route, but it is possible.  For example, you could lock on `context.RequestServices` before accessing the dbcontext.

3. Create a separate service provider scope for any code that needs to run in parallel with other code.  For example:

```cs
// this code can go in a separate class (registered as a singleton), or right in your graph class
public class MyDataClass
{
    private readonly IServiceProvider _serviceProvider;
    public MyDataClass(IServiceProvider serviceProvider)
    {
        _serviceProvider = serviceProvider;
    }

    public async Task<IEnumerable<Patient>> GetPatientsById(IEnumerable<int> ids, CancellationToken cancellationToken)
    {
        using (var scope = _serviceProvider.CreateScope())
        {
            // or, here you can call any of your other scoped services safely, regardless of how many other
           //   graph nodes are executing concurrently, either in the same request or in a separate one
            var dbContext = scope.Services.GetRequiredService<MyDbContext>();
            return await dbContext.Patients.Where(x => ids.Contains(x.Id)).ToListAsync(cancellationToken);
        }
    }
}
```

4. Write a class that copies/derives from `ParallelExecutionStrategy` with the necessary changes to run all data loaders in series.  It would be very easy; only a couple lines of code need to change.

5. I also have a library that actually supports mixing parallel and serial execution of field resolvers, with options for making a service scope if needed by a field resolver for parallel execution.  For data loaders, however, it always runs them in series.  You'd probably need to rewrite some of your graph classes to take advantage of it also.  See: https://libraries.io/nuget/Shane32.GraphQL.DI - sorry, it's only one step up from a draft, so no documentation yet.  I do use it in a production environment so it should be stable.

**Comment by grean** [comment]
Finally, i've chosen the third solution.
It will be more simultaneous connections on the DBSystem and more sql queries but we keep parallelism on.
Do you know if it´s better to do 2 sql requests with dataloader (n+1 solution problem) or a single sql request with an include statement for an SQL join ?

**Comment by grean** [comment]
I have an other question related to this problem
In the graph first level, i have an aliment that i retrieve with his id via 
```c#
 FieldAsync<AlimentType>(
                "aliment",
                description: "Retrieve an aliment with his id",
                arguments: new QueryArguments(new QueryArgument<NonNullGraphType<GuidGraphType>> { Name = "id", Description = "id of aliment to retrieve" }),
                resolve: async ctx =>
                {
                    var lol = ctx.UserContext["userId"];
                    var alimentId = ctx.GetArgument<Guid>("id");
                    return await repository.GetAlimentByIdAsync(alimentId);
                });
```
and in the repository
```c#
public async Task<Aliment> GetAlimentByIdAsync(Guid id)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var dbContext = scope.ServiceProvider.GetRequiredService<DietContext>();
                return await dbContext.Aliments.FirstAsync(i => i.Id == id);
            }
        }
```
but in the AlimentGraphqlType i have a collection of an other type that is using his parent to resolve his own props like that
```c#
//In alimentType class a have this field, that is a list of recipeAliment that are using the context aliment resolved above
            FieldAsync<ListGraphType<RecipeAlimentType>, IReadOnlyCollection<RecipeAliment>>(
                "recipeAliments",
                resolve: ctx =>
                {
                    return repository.GetRecipeAlimentsByAlimentAsync(ctx.Source); //here Source is the context Aliment
                });
```
Each recipeAliment uses an aliment dependence to calculate there props like this
```c#
    public class RecipeAliment
    {
        public Guid Id { get; set; }
        public decimal Quantity { get; set; }

        public Guid RecipeId { get; set; }
        public Recipe Recipe { get; set; }
        public Guid AlimentId { get; set; }
        public Aliment Aliment { get; set; }
        [NotMapped]
        public decimal Poids => Quantity * (Aliment.PortP ?? 0); // here, i need the aliment in the EF dbContext
        [NotMapped]
        public decimal Calories => Poids / 100 * Aliment.EKcal ?? 0;// and  here too
    }
```
But when i get all the recipeAliments, i have to do it with a new scoped dbContext because in parallel, there are others collections to fetch.
So i've passed the 'parent' aliment object to inject in the new scoped dbContext in order to resolve the dependance of my recipeAliments.
like that 
```c#
public async Task<IReadOnlyCollection<RecipeAliment>> GetRecipeAlimentsByAlimentAsync(Aliment aliment)
        //public async Task<IReadOnlyCollection<RecipeAliment>> GetRecipeAlimentsByAlimentIdAsync(Guid id)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var dbContext = scope.ServiceProvider.GetRequiredService<DietContext>();
                dbContext.Attach(aliment);
                return await dbContext.RecipeAliments.Where(i => i.AlimentId == aliment.Id).ToListAsync();
            }
        }
```
, and it works well.
But do you know if there is a better way to achieve this ? or is it the good way to do it ?
Thanks.

**Comment by Shane32** [comment]
> Do you know if it´s better to do 2 sql sequential requests with dataloader (n+1 solution problem) or a single sql request with an include statement for an SQL join ?

As far as SQL/EF, using the include statement should typically be faster with navigation properties that are not collections, such as Recipe and Aliment in the sample you posted.  I'm also sure this isn't always true -- like say you returned 1000 products, and those 1000 products had links to warehouses (of which you only had 5 warehouses), and the warehouse table had a lot of columns or perhaps a binary picture column.  Well now you are returning those 5 warehouses over and over so SQL is duplicating the warehouse data 200x over normal.  And if you use in-memory caching in your web server, perhaps those 5 warehouse rarely need to be returned.

However, if you are attempting to retrieve collection properties with the include statement, you can quickly run into the infamous 'cartesian results' problem,.....consider this sample:

- Products table
- Each product has multiple attributes unique to the product (e.g. `product.Attributes`)
- Each product has multiple categories from a common list of categories (e.g. `product.ProductCategories`, and then `productCategory.Category`)
- Each product has multiple vendors from a common list of vendors (e.g. `product.ProductVendors`, and then `productVendors.Vendor`)
- Each product has multiple pictures that are unique to the product (e.g. `product.Pictures`)

Say each product has 5 attributes, 2 categories, 2 vendors, and 5 pictures.  Then say you execute this query and retrieve 50 products:

```cs
var results = db.Products
    .Include(x => x.Attributes)
    .Include(x => x.ProductCategories).ThenInclude(x => x.Category)
    .Include(x => x.ProductVendors).ThenInclude(x => x.Vendor)
    .Include(x => x.Pictures);
```

If these are constructed with inner joins, you will be returning 5,000 records, and each record will have all the columns from all of the tables, whereas there is only 950 rows of source data across all 7 tables.  99.9% of the data returned is redundant.  It can really slow down the query execution.

EF Core 2.x and EF Core 5 can automatically split these into separate queries for you, retrieving the data across 4 database calls, and then combine the results.  (EF Core 3 cannot.)  But then you are back to your original question: is it faster to use inner join or not?  Well, not in that example.  Perhaps EF Core can do it faster than a data loader can.  The main difference is that EF Core is going to perform the original query over again, with inner joins, and just retrieve only the necessary information.  The data loader is going to queue up all the IDs that need to be retrieved and then query for those IDs.  The latter is probably faster for the SQL engine to execute, but slower for SQL to parse (since the IDs are unfortunately passed as literals).

**Comment by Shane32** [comment]
> So i've passed the 'parent' aliment object to inject in the new scoped dbContext in order to resolve the dependance of my recipeAliments.

I think you're saying that you want to be able to retrieve the recipe aliments by accessing `aliment.RecipeAliments`, which is automatically populated after the query is executed?  Then yes.  If not, then there's no real reason to attach the aliment; the database query you've written is only using `aliment.Id` which does not need to be attached.

> `.Where(i => i.AlimentId == aliment.Id)`

**Comment by grean** [comment]
Thank you for all the informations.
I only need aliment property in my recipeAliment to be filled when i access to those other properties
```c#
[NotMapped]
        public decimal Poids => Quantity * (Aliment.PortP ?? 0); // here, i need the aliment in the EF dbContext
        [NotMapped]
        public decimal Calories => Poids / 100 * Aliment.EKcal ?? 0;// and  here too
```
But as i already get the aliment object just before in the parent node, i don´t want to get it a second time with an include statement with te recipeAliments query. So i pass it through the GetRecipeAlimentsByAlimentAsync and i attach it to the new dbContext.
So you tell me yes. This is not an anti-pattern ?

**Comment by grean** [comment]
In fact, my implementation is bad because if i request all the aliments i will have the N+1 problem.
So i've changed to the case with the collection property navigation like that :
```c#
            Field<ListGraphType<RecipeAlimentType>, IEnumerable<RecipeAliment>>()
                .Name("recipeAliments")
                .ResolveAsync(ctx =>
                {
                    var recipeAlimentLoader = accessor.Context.GetOrAddCollectionBatchLoader<Guid, RecipeAliment>("GetRecipeAlimentsByAliment", repository.GetRecipeAlimentsByAlimentAsync);

                    return recipeAlimentLoader.LoadAsync(ctx.Source.Id);
                });
```
and it´s batching all the recipeAliments so it works good.
But, my dependance on each recipeAliment (Aliment property) is missing..(due to different dbContext..)
I had already resolved that problem by passing the Aliment object with my older solution, but with the loader i can´t do it anymore because i have to pass the alimentId to the loader..

Finally, the writing of this post gives me an idea but do you confirm that is not a bad idea ?
I've replaced the collect of the alimentIds by the aliment itself. I already have all those references in memory isn't it ? so i pass to the callback an `IEnumerable<Aliment> aliments` that is a collection of references as before with the collection of Guid isn't it ?
my implementation working
```c#
            Field<ListGraphType<RecipeAlimentType>, IEnumerable<RecipeAliment>>()
                .Name("recipeAliments")
                .ResolveAsync(ctx =>
                {
                    var recipeAlimentLoader = accessor.Context.GetOrAddCollectionBatchLoader<Aliment, RecipeAliment>("GetRecipeAlimentsByAliment", repository.GetRecipeAlimentsByAlimentAsync);

                    return recipeAlimentLoader.LoadAsync(ctx.Source);
                });
```
and the callback
```c#
public async Task<ILookup<Aliment, RecipeAliment>> GetRecipeAlimentsByAlimentAsync(IEnumerable<Aliment> aliments, CancellationToken cancellationToken)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var dbContext = scope.ServiceProvider.GetRequiredService<DietContext>();
                dbContext.AttachRange(aliments);
                var ids = aliments.Select(i => i.Id);
                var recipeAliments = await dbContext.RecipeAliments.Where(i => ids.Contains(i.AlimentId)).ToListAsync(cancellationToken);
                return recipeAliments.ToLookup(i => i.Aliment);
            }
        }
```

**Comment by Shane32** [comment]
My gut feeling is that you don't need to worry about attaching aliments or anything, but rather operate on the id only.  Especially if this code is used for the dataloader.

```cs
        public async Task<ILookup<Guid, RecipeAliment>> GetRecipeAlimentsByAlimentIdAsync(IEnumerable<Guid> alimentIds, CancellationToken cancellationToken)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var dbContext = scope.ServiceProvider.GetRequiredService<DietContext>();
                var recipeAliments = await dbContext.RecipeAliments.Where(i => alimentIds.Contains(i.AlimentId)).ToListAsync(cancellationToken);
                return recipeAliments.ToLookup(i => i.AlimentId);
            }
        }
```

Also maybe just return the recipes -- only if you don't need the intermediary class:

```cs
        public async Task<ILookup<Guid, Recipe>> GetRecipesByAlimentIdAsync(IEnumerable<Guid> alimentIds, CancellationToken cancellationToken)
        {
            using (var scope = _serviceProvider.CreateScope())
            {
                var dbContext = scope.ServiceProvider.GetRequiredService<DietContext>();
                var recipeAliments = await dbContext.RecipeAliments
                    .Where(i => alimentIds.Contains(i.AlimentId))
                    .Select(i => new { AlimentId = i.AlimentId, i.Recipe })
                    .ToListAsync(cancellationToken);
                return recipeAliments.ToLookup(i => i.AlimentId, i => i.Recipe);
            }
        }
```

**Comment by Shane32** [comment]
> Finally, the writing of this post gives me an idea but do you confirm that is not a bad idea ?
> I've replaced the collect of the alimentIds by the aliment itself. I already have all those references in memory isn't it ? so i pass to the callback an `IEnumerable<Aliment> aliments` that is a collection of references as before with the collection of Guid isn't it ?

I dunno; I guess you can.  I'm not sure why you'd want to, but if that's what works for your needs....

**Comment by grean** [comment]
I want Aliments entities to be in the dbContext because in the alimentRecipe class, I need some information provided by the Food navigation property. This is useful for performing calculations such as the amount of KiloCalories that a food can add to a recipe for example. I know that i can do this calcul client side, but i'm pretty sur that there are others use cases that need to resolve navigation properties dependencies don´t you think ?
```c#
    public class RecipeAliment
    {
        public Guid Id { get; set; }
        public decimal Quantity { get; set; }

        public Guid RecipeId { get; set; }
        public Recipe Recipe { get; set; }
        public Guid AlimentId { get; set; }
        public Aliment Aliment { get; set; }

        [NotMapped]
        public decimal Poids => Quantity * (Aliment.PortP ?? 0);
        [NotMapped]
        public decimal Calories => Poids / 100 * Aliment.EKcal ?? 0;
    }
```
I've made some tests on my machine to my api and when i use the 'normal way' of getting the GUID, i can tell that it´s slower than attaching all the entities (i test with 2500 aliment in my DB)because of the SQL join statement that i have to do
```c#
var recipeAliments = await dbContext.RecipeAliments.Include(i => i.Aliment).Where(i => ids.Contains(i.AlimentId)).ToListAsync(cancellationToken);
```
The gain is a third and attaching 2500 objects to the context is very fast.

**Comment by Shane32** [comment]
so instead of doing 
```cs
var recipeAliments = await dbContext.RecipeAliments.Include(i => i.Aliment).Where(i => ids.Contains(i.AlimentId)).ToListAsync(cancellationToken);
```
you might be able to do this without using dataloaders or having to use Attach:
```cs
var recipeAliments = await dbContext.RecipeAliments.Where(i => ids.Contains(i.AlimentId)).ToListAsync(cancellationToken);
await dbContext.RecipeAliments.Where(i => ids.Contains(i.AlimentId)).SelectMany(i => i.Aliment).LoadAsync(cancellationToken);
```
Just a thought.  But it sounds like you've got it pretty optimized for your needs.
